x-pack/plugins/security/server/authentication/authentication_service.ts

```diff
@@ -139,7 +155,43 @@ export class AuthenticationService {
       return t.notHandled();
     });
 
-    this.logger.debug('Successfully registered core authentication handler.');
+    http.registerOnPreResponse((request, preResponse, toolkit) => {
+      if (preResponse.statusCode !== 401 || !canRedirectRequest(request)) {
+        return toolkit.next();
+      }
+
+      if (!this.authenticator) {
+        // Core doesn't allow returning error here.
+        this.logger.error('Authentication sub-system is not fully initialized yet.');
+        return toolkit.next();
+      }
+
+      // If users can eventually re-login we want to redirect them directly to the page they tried
+      // to access initially, but we only want to do that for non-API routes. API routes that support
+      // redirects are solely used to support various authentication flows that wouldn't make any
+      // sense after successful authentication through login page.
+      const originalURL = !API_ROUTES_SUPPORTING_REDIRECTS.includes(request.route.path)
```

Comment:
> API routes that support redirects are solely used to support various authentication flows that wouldn't make any sense after successful authentication through login page.

If we ignore reporting (see other comment), then this is true today, but might not always be true. It might be worth either:
1) Renaming this constant to make the intent clear (only used for auth flows that don't make sense after authentication)
2) Declare another constant with a clearer name, which initially the same values as `API_ROUTES_SUPPORTING_REDIRECTS`. This would allow the two sets to evolve over time for different reasons, without introducing subtle bugs along the way

What do you think?

Reply:
Yeah, that's a fair point. If we decide to support reporting endpoints I'll pick the second option, otherwise the first one and hopefully we can extend Core router to get rid of these hardcoded lists.